File: api/app/models/theq/office.py
'''Copyright 2018 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.'''

from qsystem import db
from app.models.theq import Base
from app.models.bookings import Exam, Room
from qsystem import cache, db
import enum
from sqlalchemy import Enum, desc


class Status(enum.Enum):
    SHOW = 'SHOW'
    HIDE = 'HIDE'
    DISABLE = 'DISABLE'


class Office(Base):
    # Defining String constants to appease SonarQube
    office_id_const = 'office.office_id'
    service_id_const = 'service.service_id'
    
    office_service = db.Table(
        'office_service',
        db.Column('office_id', db.Integer,
                db.ForeignKey(office_id_const, ondelete="CASCADE"), primary_key=True),
        db.Column('service_id', db.Integer,
                db.ForeignKey(service_id_const, ondelete="CASCADE"), primary_key=True))

    office_quick_list = db.Table(
        'office_quick_list',
        db.Column('office_id', db.Integer,
                db.ForeignKey(office_id_const, ondelete="CASCADE"), primary_key=True),
        db.Column('service_id', db.Integer,
                db.ForeignKey(service_id_const, ondelete="CASCADE"), primary_key=True))

    office_back_office_list = db.Table(
        'office_back_office_list',
        db.Column('office_id', db.Integer,
                db.ForeignKey(office_id_const, ondelete="CASCADE"), primary_key=True),
        db.Column('service_id', db.Integer,
                db.ForeignKey(service_id_const, ondelete="CASCADE"), primary_key=True))

    office_counter= db.Table(
        'office_counter',
        db.Column('office_id', db.Integer,
                db.ForeignKey(office_id_const, ondelete="CASCADE"), primary_key=True),
        db.Column('counter_id', db.Integer,
                db.ForeignKey('counter.counter_id', ondelete="CASCADE"), primary_key=True))

    office_timeslot = db.Table(
        'office_timeslot',
        db.Column('office_id', db.Integer,
                  db.ForeignKey(office_id_const, ondelete="CASCADE"), primary_key=True),
        db.Column('time_slot_id', db.Integer,
                  db.ForeignKey('timeslot.time_slot_id', ondelete="CASCADE"), primary_key=True))

    office_id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    office_name = db.Column(db.String(100))
    office_number = db.Column(db.Integer)
    sb_id = db.Column(db.Integer, db.ForeignKey('smartboard.sb_id'))
    deleted = db.Column(db.DateTime, nullable=True)
    exams_enabled_ind = db.Column(db.Integer, nullable=False)
    appointments_enabled_ind = db.Column(db.Integer, nullable=False, default=0)
    timezone_id = db.Column(db.Integer, db.ForeignKey('timezone.timezone_id'), nullable=True)
    latitude = db.Column(db.Float)
    longitude = db.Column(db.Float)
    office_appointment_message = db.Column(db.String(1000))
    appointments_days_limit = db.Column(db.Integer, default=30)
    appointment_duration = db.Column(db.Integer, default=30)
    max_person_appointment_per_day = db.Column(db.Integer, default=1)
    civic_address = db.Column(db.String(200))
    telephone = db.Column(db.String(20))
    online_status = db.Column(Enum(Status))
    number_of_dlkt = db.Column(db.Integer, nullable=True)
    office_email_paragraph = db.Column(db.String(2000), nullable=True)
    external_map_link = db.Column(db.String(500), nullable=True)
    soonest_appointment = db.Column(db.Integer, default=0)

    counters = db.relationship("Counter", secondary='office_counter')
    services = db.relationship("Service", secondary='office_service')
    quick_list = db.relationship("Service", secondary='office_quick_list')
    back_office_list = db.relationship("Service", secondary='office_back_office_list')
    csrs = db.relationship('CSR')
    citizens = db.relationship('Citizen', backref='office_citizens')
    timeslots = db.relationship('TimeSlot')

    sb = db.relationship('SmartBoard')
    timezone = db.relationship('Timezone')

    exams = db.relationship("Exam")
    rooms = db.relationship('Room')

    # for walk-in notifications
    check_in_notification = db.Column(db.Integer)
    check_in_reminder_msg = db.Column(db.Text)
    automatic_reminder_at = db.Column(db.Integer)
    # for Digital Signage
    currently_waiting = db.Column(db.Integer)
    digital_signage_message = db.Column(db.Integer)
    digital_signage_message_1 = db.Column(db.Text)
    digital_signage_message_2 = db.Column(db.Text)
    digital_signage_message_3 = db.Column(db.Text)
    show_currently_waiting_bottom = db.Column(db.Integer)
    

    format_string = 'office_%s'
    offices_cache_key: str = 'active_offices'

    def __repr__(self):
        return self.office_name

    def __init__(self, **kwargs):
        super(Office, self).__init__(**kwargs)

    @classmethod
    def find_by_id(cls, office_id: int):
        """Return a Office by office_id."""
        key = Office.format_string % office_id
        office = cache.get(key)
        if not office:
            office = cls.query.get(office_id)
            office.timeslots
            office.timezone
        return office

    @classmethod
    def build_cache(cls):
        """Build cache."""
        try:
            all_offices = cls.query.all()
            for office in all_offices:
                key = Office.format_string % office.office_id
                office.timeslots
                office.timezone
                cache.set(key, office)
        except Exception:
            print('Error on building cache')

    @classmethod
    def get_all_active_offices(cls):
        """Return all active offices."""
        from app.schemas.theq import OfficeSchema

        active_offices = cache.get(Office.offices_cache_key)
        if not active_offices:
            office_schema = OfficeSchema(many=True)
            active_offices = office_schema.dump(Office.query.filter(Office.deleted.is_(None)).order_by(Office.office_name))
            cache.set(Office.offices_cache_key, active_offices)
        return active_offices

    @classmethod
    def clear_offices_cache(cls):
        """Clear active offices cache."""
        cache.delete(Office.offices_cache_key)

File: api/static/js/office.js
$(document).ready(function() {
    // Smartboard
    var sb = $("#sb"); 
    if (sb.val() === "__None" || sb.val() === "3") { // if not selected or nocallonsmartboard
        toggleSmartboardFields(false);
    }
    sb.on('change', (e) => {        
        if (e.val === "__None" || e.val === "3") { // if not selected or nocallonsmartboard
            toggleSmartboardFields(false);     
        } else { // if selected and NOT nocallonsmartboard            
            toggleSmartboardFields(true);
        }
    })

    // Digital Signage
    var digital_signage_message = $("#digital_signage_message");
    if (digital_signage_message.val() === "__None" || digital_signage_message.val() === "0") { // if not selected or disabled
        toggleDigitalSignageFields(false);
    }
    digital_signage_message.on('change', (e) => {
        if (e.val  === "__None" || e.val === "0") { // if not selected or disabled
            toggleDigitalSignageFields(false);
        } else { // if enabled
            toggleDigitalSignageFields(true);
        }
    });

    // Check-In Notifications
    var check_in_notification = $("#check_in_notification");
    if (check_in_notification.val() === "__None" || check_in_notification.val() === "0") { // if not selected or disabled
        toggleCheckInNotificationFields(false);
    }
    check_in_notification.on('change', (e) => { 
        if (e.val === "__None" || e.val === "0") { // if not selected or disabled
            toggleCheckInNotificationFields(false);
        } else { // if enabled
            toggleCheckInNotificationFields(true);
        }
    });
    
    // Appointments Enabled
    var appointments_enabled_ind = $("#appointments_enabled_ind");
    if (appointments_enabled_ind.val() === "0") { // if disabled
        toggleAppointmentFields(false);
        toggleOnlineStatusFields(false);
    }        
    appointments_enabled_ind.on('change', (e) => {        
        if (e.val === "0") { // if disabled            
            toggleAppointmentFields(false);
            toggleOnlineStatusFields(false);
        } else { // if enabled            
            toggleAppointmentFields(true);
        }
    });
    
    // Online Status
    var online_status = $("#online_status");
    if (online_status.val() === "__None" || online_status.val() === "HIDE") { // if not selected or HIDE
        toggleOnlineStatusFields(false);
    }        
    online_status.on('change', (e) => {        
        if (e.val === "__None" || e.val === "HIDE") { // if not selected or HIDE
            toggleOnlineStatusFields(false);
        } else { // if SHOW
            if (appointments_enabled_ind.val() === "0") { // if appointment is disabled, hide Online Status fields still
                toggleOnlineStatusFields(false);
            } else { // if appointment is enabled, show Online Status fields
                toggleOnlineStatusFields(true);
            }            
        }
    });
});

function toggleParentContainer(el, flag) {
    if (flag)
        el.parent().show();
    else
        el.parent().hide();
}

function toggleParentContainers(arr, flag) {
    arr.forEach((el, index) => {  
        toggleParentContainer(el, flag);
    });
}

function toggleSmartboardFields(flag) {
    var arr = [];
    arr.push($("label[for='currently_waiting']"));
    arr.push($("label[for='show_currently_waiting_bottom']"));
    toggleParentContainers(arr, flag);
}

function toggleDigitalSignageFields(flag) {
    var arr = [];
    arr.push($("label[for='digital_signage_message_1']"));
    arr.push($("label[for='digital_signage_message_2']"));
    arr.push($("label[for='digital_signage_message_3']"));
    toggleParentContainers(arr, flag);
}

function toggleCheckInNotificationFields(flag) {
    var arr = [];
    arr.push($("label[for='check_in_reminder_msg']"));
    arr.push($("label[for='automatic_reminder_at']"));
    toggleParentContainers(arr, flag);
}

function toggleAppointmentFields(flag) {
    var arr = [];
    arr.push($("label[for='appointment_duration']"));
    arr.push($("label[for='office_email_paragraph']"));
    arr.push($("label[for='online_status']"));
    toggleParentContainers(arr, flag);
}

function toggleOnlineStatusFields(flag) {
    var arr = [];
    arr.push($("label[for='office_appointment_message']"));
    arr.push($("label[for='latitude']"));
    arr.push($("label[for='longitude']"));
    arr.push($("label[for='civic_address']"));
    arr.push($("label[for='telephone']"));
    arr.push($("label[for='external_map_link']"));
    arr.push($("label[for='appointments_days_limit']"));
    arr.push($("label[for='soonest_appointment']"));
    arr.push($("label[for='max_person_appointment_per_day']"));
    arr.push($("label[for='number_of_dlkt']"));
    arr.push($("label[for='timeslots']"));
    toggleParentContainers(arr, flag);
}